keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,336 @@
+type SELinuxRelabelPolicy string
+
+const (
+    Mount         SELinuxRelabelPolicy = "Mount"
+    AlwaysRelabel SELinuxRelabelPolicy = "Always"
+)
+
+type PodSecurityContext `struct {
+    // SELinuxRelabelPolicy ← new field
+    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
+    // Valid values are "Mount" and "Always". If not specified, "Always" is used.
+    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
+    // "Mount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+    // change.
+    // + optional
+    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+
+    // For context:
+    // fsGroupChangePolicy defines behavior of changing ownership and permission of the volume
+    // before being exposed inside Pod. This field will only apply to
+    // volume types which support fsGroup based ownership(and permissions).
+    // It will have no effect on ephemeral volume types such as: secret, configmaps
+    // and emptydir.
+    // Valid values are "OnRootMismatch" and "Always". If not specified defaults to "Always".
+    // +optional
+    FSGroupChangePolicy *PodFSGroupChangePolicy `json:"fsGroupChangePolicy,omitempty" protobuf:"bytes,9,opt,name=fsGroupChangePolicy"`
+
+    ...
+}
+```
+
+See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20200120-skip-permission-change.md for similar API for ownership change for fsGroup.
+This KEP should follow API provided for fsGroup closely, however, the implementation is different (mount vs. recursive `chown`).
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### `mount -o context`
+Linux kernel, with SELinux compiled in, allows `mount -o context=s0:c309,c383 <what> <where>` to mount a volume and pretend that all files on the volume have given SELinux label.
+It works only for the first mount of the volume!
+It does not work for bind-mounts or any subsequent mount of the same volume.
+
+### New Kubernetes behavior
+
+* When kubelet *knows* SELinux context of a pod / container to run (i.e. Pod/Container contains `SELinuxOptions`) and `SELinuxRelabelPolicy` is `Mount`, it tries to mount all volumes for the Pod with given SELinux label using `mount -o context=XYZ`.
+    Kubelet makes sure that the option is passed to the first mount in all in-tree volume plugins (incl. ephemeral volumes like Secrets).
+    Kubelet passes it as an mount option to all CSI calls for given volume.
+
+    After the volume is mounted, kubelet checks that the root of the volume has the expected SELinux label, i.e. that the volume was mounted correctly.
+    * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
+    * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly or the volume was already mounted:
```

Comment:
When we use `-o context` do we also need to specify user, role fields? when CRI does chcon it seems to default to values defined by OS/CRI configuration.

Reply:
Yes, `-o context` must contain user and role, otherwise mount fails with `wrong fs type, bad option, bad superblock on xxx`. I don't see a way how kubelet could get the defaults from anywhere. I added it as "Open Question".

Reply:
How are we going to verify if volume has correct label if we can't query user/role etc from kubelet?

Reply:
Added a note for that too.